azurerm/resource_arm_monitor_metric_alert.go

```diff
@@ -0,0 +1,450 @@
+			Enabled:             utils.Bool(enabled),
+			Description:         utils.String(description),
+			Severity:            utils.Int32(int32(severity)),
+			EvaluationFrequency: utils.String(frequency),
+			WindowSize:          utils.String(windowSize),
+			Scopes:              expandMonitorMetricAlertStringArray(targetsRaw),
+			Criteria:            expandMonitorMetricAlertCriteria(criteriaRaw),
+			Actions:             expandMonitorMetricAlertAction(actionRaw),
+		},
+		Tags: expandedTags,
+	}
+
+	if _, err := client.CreateOrUpdate(ctx, resGroup, name, parameters); err != nil {
+		return fmt.Errorf("Error creating or updating metric alert %q (resource group %q): %+v", name, resGroup, err)
+	}
+
+	read, err := client.Get(ctx, resGroup, name)
+	if err != nil {
+		return err
+	}
+	if read.ID == nil {
+		return fmt.Errorf("Metric alert %q (resource group %q) ID is empty", name, resGroup)
+	}
+	d.SetId(*read.ID)
+
+	return resourceArmMonitorMetricAlertRead(d, meta)
+}
+
+func resourceArmMonitorMetricAlertRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).monitorMetricAlertsClient
+	ctx := meta.(*ArmClient).StopContext
+
+	id, err := parseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resGroup := id.ResourceGroup
+	name := id.Path["metricAlerts"]
+
+	resp, err := client.Get(ctx, resGroup, name)
+	if err != nil {
+		if response.WasNotFound(resp.Response.Response) {
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Error getting metric alert %q (resource group %q): %+v", name, resGroup, err)
+	}
+
+	d.Set("name", name)
+	d.Set("resource_group_name", resGroup)
+	if alert := resp.MetricAlertProperties; alert != nil {
+		d.Set("enabled", alert.Enabled)
+		d.Set("description", alert.Description)
+		d.Set("severity", alert.Severity)
+		d.Set("frequency", alert.EvaluationFrequency)
+		d.Set("window_size", alert.WindowSize)
+		if err := d.Set("target_ids", flattenMonitorMetricAlertStringArray(alert.Scopes)); err != nil {
+			return err
+		}
+		if err := d.Set("criteria", flattenMonitorMetricAlertCriteria(alert.Criteria)); err != nil {
+			return err
+		}
+		if err := d.Set("action", flattenMonitorMetricAlertAction(alert.Actions)); err != nil {
+			return err
+		}
+	}
+	flattenAndSetTags(d, resp.Tags)
+
+	return nil
+}
+
+func resourceArmMonitorMetricAlertDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).monitorMetricAlertsClient
+	ctx := meta.(*ArmClient).StopContext
+
+	id, err := parseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resGroup := id.ResourceGroup
+	name := id.Path["metricAlerts"]
+
+	if resp, err := client.Delete(ctx, resGroup, name); err != nil {
+		if !response.WasNotFound(resp.Response) {
+			return fmt.Errorf("Error deleting metric alert %q (resource group %q): %+v", name, resGroup, err)
+		}
+	}
+
+	return nil
+}
+
+func expandMonitorMetricAlertStringArray(v []interface{}) *[]string {
```

Comment:
This is a pretty common piece of code, might be work moving into utils?

Reply:
I agree. But I'm not sure whether to put it into `utils` or `helpers`?